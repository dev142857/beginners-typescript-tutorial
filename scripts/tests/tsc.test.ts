import fg from "fast-glob";
import path from "path";
import { describe, expect, it } from "vitest";
import { execSync } from "child_process";

const srcFolder = path.resolve(__dirname, "../../src");

const allFiles = path.resolve(srcFolder, "*{problem,solution,explainer}*.ts");

const toRelativeFilename = (file: string) => {
  return path.relative(srcFolder, file);
};

describe("tsc", async () => {
  const files = await fg(allFiles);

  it.each(files.map(toRelativeFilename))(
    "In file %s, TS errors should match snapshot",
    (file) => {
      const resolvedFile = path.join(srcFolder, file);
      let result: string;

      try {
        result = execSync(`tsc "${resolvedFile}" --noEmit --strict`).toString();
      } catch (error) {
        result = error.output.toString();
      }

      expect(result).toMatchSnapshot();
    },
  );
});
