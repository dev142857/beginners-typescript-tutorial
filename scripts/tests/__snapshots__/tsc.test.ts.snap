// Vitest Snapshot v1

exports[`tsc > In file 01-number.problem.ts, TS errors should match snapshot 1`] = `
",../../src/01-number.problem.ts(3,31): error TS7006: Parameter 'a' implicitly has an 'any' type.
../../src/01-number.problem.ts(3,34): error TS7006: Parameter 'b' implicitly has an 'any' type.
,"
`;

exports[`tsc > In file 01-number.solution.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 02-object-param.problem.ts, TS errors should match snapshot 1`] = `
",../../src/02-object-param.problem.ts(3,31): error TS7006: Parameter 'params' implicitly has an 'any' type.
,"
`;

exports[`tsc > In file 02-object-param.solution.1.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 02-object-param.solution.2.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 02-object-param.solution.3.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 03-optional-properties.problem.ts, TS errors should match snapshot 1`] = `
",../../src/03-optional-properties.problem.ts(11,24): error TS2345: Argument of type '{ first: string; }' is not assignable to parameter of type '{ first: string; last: string; }'.
  Property 'last' is missing in type '{ first: string; }' but required in type '{ first: string; last: string; }'.
,"
`;

exports[`tsc > In file 03-optional-properties.solution.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 04-optional-params.problem.ts, TS errors should match snapshot 1`] = `
",../../src/04-optional-params.problem.ts(11,16): error TS2554: Expected 2 arguments, but got 1.
,"
`;

exports[`tsc > In file 04-optional-params.solution.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 05-assigning-types-to-variables.problem.ts, TS errors should match snapshot 1`] = `
",../../src/05-assigning-types-to-variables.problem.ts(21,20): error TS2345: Argument of type '{}' is not assignable to parameter of type 'User'.
  Type '{}' is missing the following properties from type 'User': id, firstName, lastName, isAdmin
,"
`;

exports[`tsc > In file 05-assigning-types-to-variables.solution.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 06-unions.problem.ts, TS errors should match snapshot 1`] = `
",../../src/06-unions.problem.ts(18,3): error TS2578: Unused '@ts-expect-error' directive.
,"
`;

exports[`tsc > In file 06-unions.solution.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 07-arrays.problem.ts, TS errors should match snapshot 1`] = `
",../../src/07-arrays.problem.ts(19,3): error TS2739: Type '{ id: number; title: string; }[]' is missing the following properties from type 'Post': id, title
,"
`;

exports[`tsc > In file 07-arrays.solution.1.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 07-arrays.solution.2.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 08-function-return-type-annotations.problem.ts, TS errors should match snapshot 1`] = `
",../../src/08-function-return-type-annotations.problem.ts(27,15): error TS2339: Property 'id' does not exist on type '{}'.
../../src/08-function-return-type-annotations.problem.ts(28,15): error TS2339: Property 'firstName' does not exist on type '{}'.
../../src/08-function-return-type-annotations.problem.ts(29,15): error TS2339: Property 'lastName' does not exist on type '{}'.
../../src/08-function-return-type-annotations.problem.ts(30,15): error TS2339: Property 'role' does not exist on type '{}'.
../../src/08-function-return-type-annotations.problem.ts(32,15): error TS2339: Property 'posts' does not exist on type '{}'.
../../src/08-function-return-type-annotations.problem.ts(33,15): error TS2339: Property 'posts' does not exist on type '{}'.
,"
`;

exports[`tsc > In file 08-function-return-type-annotations.solution.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 09-promises.problem.ts, TS errors should match snapshot 1`] = `
",../../src/09-promises.problem.ts(12,45): error TS1055: Type 'LukeSkywalker' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.
,"
`;

exports[`tsc > In file 09-promises.solution.1.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 09-promises.solution.2.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 09-promises.solution.3.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 10-set.problem.ts, TS errors should match snapshot 1`] = `
",../../src/10-set.problem.ts(15,3): error TS2578: Unused '@ts-expect-error' directive.
../../src/10-set.problem.ts(22,24): error TS2344: Type 'false' does not satisfy the constraint 'true'.
,"
`;

exports[`tsc > In file 10-set.solution.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 11-record.problem.ts, TS errors should match snapshot 1`] = `
",../../src/11-record.problem.ts(7,5): error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{}'.
  No index signature with a parameter of type 'string' was found on type '{}'.
../../src/11-record.problem.ts(11,12): error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{}'.
  No index signature with a parameter of type 'string' was found on type '{}'.
../../src/11-record.problem.ts(26,10): error TS7053: Element implicitly has an 'any' type because expression of type '\\"123\\"' can't be used to index type '{}'.
  Property '123' does not exist on type '{}'.
../../src/11-record.problem.ts(35,10): error TS7053: Element implicitly has an 'any' type because expression of type '\\"123\\"' can't be used to index type '{}'.
  Property '123' does not exist on type '{}'.
,"
`;

exports[`tsc > In file 11-record.solution.1.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 11-record.solution.2.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 11-record.solution.3.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 12-typeof-narrowing.problem.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 12-typeof-narrowing.solution.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 13-catch-blocks.problem.ts, TS errors should match snapshot 1`] = `
",../../src/13-catch-blocks.problem.ts(9,12): error TS2571: Object is of type 'unknown'.
,"
`;

exports[`tsc > In file 13-catch-blocks.solution.1.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 13-catch-blocks.solution.2.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 13-catch-blocks.solution.3.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 14-extends.problem.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 14-extends.solution.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 15-intersection.problem.ts, TS errors should match snapshot 1`] = `
",../../src/15-intersection.problem.ts(34,13): error TS2571: Object is of type 'unknown'.
,"
`;

exports[`tsc > In file 15-intersection.solution.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 16-omit-and-pick.problem.ts, TS errors should match snapshot 1`] = `
",../../src/16-omit-and-pick.problem.ts(16,22): error TS2344: Type 'false' does not satisfy the constraint 'true'.
,"
`;

exports[`tsc > In file 16-omit-and-pick.solution.1.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 16-omit-and-pick.solution.2.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 17-function-types.problem.ts, TS errors should match snapshot 1`] = `
",../../src/17-function-types.problem.ts(8,5): error TS2571: Object is of type 'unknown'.
../../src/17-function-types.problem.ts(12,5): error TS2571: Object is of type 'unknown'.
../../src/17-function-types.problem.ts(16,14): error TS7006: Parameter 'isFocused' implicitly has an 'any' type.
../../src/17-function-types.problem.ts(19,24): error TS2344: Type 'false' does not satisfy the constraint 'true'.
,"
`;

exports[`tsc > In file 17-function-types.solution.1.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 17-function-types.solution.2.ts, TS errors should match snapshot 1`] = `""`;

exports[`tsc > In file 18-function-types-with-promises.problem.ts, TS errors should match snapshot 1`] = `
",../../src/18-function-types-with-promises.problem.ts(13,32): error TS2571: Object is of type 'unknown'.
../../src/18-function-types-with-promises.problem.ts(15,22): error TS2571: Object is of type 'unknown'.
../../src/18-function-types-with-promises.problem.ts(23,12): error TS7006: Parameter 'id' implicitly has an 'any' type.
,"
`;

exports[`tsc > In file 18-function-types-with-promises.solution.ts, TS errors should match snapshot 1`] = `""`;
